### DISCUSSION TITLE
Can pyinstaller work with any recent version of ubuntu ?

### DISCUSSION THREAD

**Discussion by peter1john** [pyinstaller/pyinstaller#7925] [PyInstaller]
HI,

I am new to this and while exploring how to create executable for python script i found pyinstaller is very popular and from command line installed this package and started creating executable it is working fine on my laptop. But while going throgh documentaion i found below infromation, if any of below three situation comes then we need build for that platform,

a different OS

a different version of Python

a 32-bit or 64-bit OS

My aim is to create executable for atleast 18.04 and 20.04 with any of python package.

Is there any way to crack this? 

Thank You.

**Comment by bwoodsend** [comment]
See [docs](https://pyinstaller.org/en/stable/usage.html#making-gnu-linux-apps-forward-compatible). tl;dr is that if you build on 18.04 it will usually work on 20.04 whereas it will generally not work the other way around. It does depend on the complexity of your dependencies and how many versions you consider *recent*. There are always exceptions though -- I vaguely recall it being impossible to support many versions in one build when PyQt is a dependency for example.

**Reply by bwoodsend** [reply]
Probably also worth bearing in mind that Linux on `aarch64` is quite popular now whereas `i386` (a.k.a 32 bit) is mostly dead.

**Reply by peter1john** [reply]
Yes, but in my case I just tried to build it on 20.04 and then cross checked on 18.04. 
On all 20.04 machine it is working fine without any issue.
But on when I run on 18.04 machine got below error,

[] error in loading Python lib 'tmp/_xxxxxx/linpython3.8.so.10': dlopen: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29` not found ( required by 'tmp/_xxxxxx/linpython3.8.so.10'

Any way to resolve this issue ?

**Reply by peter1john** [reply]
Yeah, please ignore 32 bit. It is more than enough if it works on 64 bit machine.

**Reply by rokm** [reply]
> Any way to resolve this issue ?

No - a shared library that was compiled against a certain version of `glibc` (2.31 in Ubuntu 20.04) cannot be loaded on a system with older version of `glibc` (2.27 on Ubuntu 18.04). Therefore, a program built on Ubuntu 20.04 cannot run on 18.04. This is not specific to PyInstaller, but to linux binaries with dynamic library dependencies in general. 

Depending on how complex your dependencies are, you might be able to work around this with a tool like `staticx`. But in general, you should either build for each distribution (and each distribution version) separately, or build on the distribution with the oldest version of `glibc` you are planning to support (but even then, there is no guarantee of compatibility across different distributions, or even across different versions of the same distribution - it all depends on your dependencies and their ABI changes).

**Reply by bwoodsend** [reply]
> Yes, but in my case I just tried to build it on 20.04 and then cross checked on 18.04.

Uhm, that is the exact opposite of what I told you would work.

**Reply by peter1john** [reply]
I have only one main python file, which included around 4-5 python files internally. So could you please give more information about staticx ?

**Reply by peter1john** [reply]
Sorry for confusion, but i need it to work on all ubuntu version :)

**Reply by bwoodsend** [reply]
> I have only one main python file, which included around 4-5 python files internally. So could you please give more information about staticx ?

But presumably there are dependencies involved too? Large, compiled libraries like numpy/Qt/Gtk are what makes the biggest differences.

> So could you please give more information about staticx ?

Are [their docs](https://staticx.readthedocs.io/en/latest/index.html) not enough?

**Reply by peter1john** [reply]
But presumably there are dependencies involved too? Large, compiled libraries like numpy/Qt/Gtk are what makes the biggest differences. ---- > Actually i am not using numpy/Qt/Gtk  packages in our script. Is it by default included in pyinstaller package ?

Are [their docs](https://staticx.readthedocs.io/en/latest/index.html) not enough?  ----- > I could see below is command to work with.

staticx -l /path/to/fancy/library.so /path/to/exe /path/to/output

Here- 
/path/to/fancy/library.so - New shared library which i need include.
 /path/to/exe  - existing executbale
/path/to/output - output path

But when i try with above options, i got below error,

Assertion failed: splitIndex != -1 (patchelf.cc: shiftFile: 504)
staticx: patchelf returned -6

**Reply by rokm** [reply]
If you don't have complex dependencies, why not build on the oldest version of Ubuntu that you plan to support?

**Reply by peter1john** [reply]
For me from 16.04 till 20.04 is sufficient. But in 16.04 and 18.04 python3.6 will come when we try to install with apt. But with 3.6 having some issues with pyinstaller. Appears to be some dependency issue.

**Reply by rokm** [reply]
> But in 16.04 and 18.04 python3.6 will come when we try to install with apt.

Then build a supported python version from source and build your application using that?

---

If you want to use `staticx`, you will probably need to make a `onefile` executable, and process that using `staticx`. I don't think it will work with `onedir` builds, because while `-l` flag allows you to bundle additional shared libs, there is no equivalent option for data files (whereas `staticx` natively supports PyInstaller's `onefile` executables).

**Reply by bwoodsend** [reply]
You can get sensible versions of Python on old Ubuntu verisons via the deadsnakes PPA https://github.com/pyinstaller/pyinstaller/issues/6358#issuecomment-963171172

**Reply by peter1john** [reply]
@rokm I have tried onefile executable itself.....

**Reply by peter1john** [reply]
@bwoodsend  Thank you, let me try this approach.

**Reply by peter1john** [reply]
@bwoodsend  - While running apt-get install -y python3.9 python3.9-dev python3.9-venv command got below error,

E: Unable to locate package python3.9

Note: I am not using any of virtual machine/docker/venv. Simple ubuntu 18.04 OS using.

**Reply by bwoodsend** [reply]
Presumably you didn't run `add-apt-repository ppa:deadsnakes/ppa` or `apt-get update`.

**Reply by peter1john** [reply]
@bwoodsend i tried both steps, appears to be issue with pip package.
resolved with help of [https://github.com/pypa/pip/issues/5599](url)

**Reply by peter1john** [reply]
Is it possible to statically link with glibc while creating executable, by assuming it wont be an issue regardless of ubuntu version as it will be part of executable itself ?. Is my understanding correct ?

**Reply by rokm** [reply]
That's the idea behind static linking, yes. 

However, PyInstaller does not *compile* shared libraries and other binaries - it just collects them from the build system. So even if you statically compiled PyInstaller's bootloader (instead of using the pre-built one), you would also need to somehow ensure that your whole python environment was built statically. 

`staticx` emulates static linking for binaries that are originally dynamically linked, so I'd say that's pretty much your only option in this regard.

**Reply by peter1john** [reply]
Thank You @rokm, let me explore bit more on it.

**Reply by peter1john** [reply]
But now this issue resolved, as mentioned earlier i tried to compile on lower version of ubuntu which we are going to support and it is working on all higher versions of ubuntu. Thank You all for helping me to resolve this issue :)
